### PR TITLE
Confirm cancel behaviour in test

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

**Comment by yschimke** [comment]
https://github.com/square/okhttp/issues/1902

File: okhttp/src/test/java/okhttp3/internal/http/CancelTest.java [hunk 1]
```diff
@@ -32,26 +33,48 @@
...
+  enum CancelMode {
+    CANCEL,
+    INTERRUPT
+  }
+
   @Rule public final OkHttpClientTestRule clientTestRule = new OkHttpClientTestRule();
+  private final CancelMode mode;
+
+  @Parameterized.Parameters(name = "{0}")
+  public static Collection<CancelMode> createStories() {
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/internal/http/CancelTest.java:63]: oooh nice method name

**Comment by swankjesse** [comment]
```

okhttp3.internal.http.CancelTest > cancelWritingRequestBody[INTERRUPT] FAILED

    java.lang.InterruptedException

        at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1367)

        at java.base/java.util.concurrent.CountDownLatch.await(CountDownLatch.java:278)

        at okhttp3.OkHttpClientTestRule.ensureAllTaskQueuesIdle(OkHttpClientTestRule.kt:99)

        at okhttp3.OkHttpClientTestRule.access$ensureAllTaskQueuesIdle(OkHttpClientTestRule.kt:36)

        at okhttp3.OkHttpClientTestRule$apply$1.evaluate(OkHttpClientTestRule.kt:129)
```

```

okhttp3.internal.http.CancelTest > cancelReadingResponseBody[INTERRUPT] FAILED

    java.lang.InterruptedException

        at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1367)

        at java.base/java.util.concurrent.CountDownLatch.await(CountDownLatch.java:278)

        at okhttp3.OkHttpClientTestRule.ensureAllTaskQueuesIdle(OkHttpClientTestRule.kt:99)

        at okhttp3.OkHttpClientTestRule.access$ensureAllTaskQueuesIdle(OkHttpClientTestRule.kt:36)

        at okhttp3.OkHttpClientTestRule$apply$1.evaluate(OkHttpClientTestRule.kt:129)
```

**Comment by swankjesse** [comment]
I think you just need to put a check to `Thread.interrupted()` somewhere to clear the thread’s interrupted status.

File: okhttp/src/test/java/okhttp3/internal/http/CancelTest.java [hunk 2]
```diff
@@ -101,6 +128,7 @@
       call.execute();
       fail();
     } catch (IOException expected) {
+      assertEquals(mode == CancelMode.INTERRUPT, Thread.interrupted());
```
Comment (yschimke) [inline okhttp/src/test/java/okhttp3/internal/http/CancelTest.java:131]: @swankjesse Good call, I think this is useful to test.

**Pull request merged**